File: Awesome.AI.Source/Awesome.AI/Common/GPTVector2D.cs
namespace Awesome.AI.Common
{
    public struct GPTVector2D
    {
        public double xx { get; set; }
        public double yy { get; set; }

        public double theta_in_radians { get; set; }
        public double theta_in_degrees { get { return ToDegrees(this); } }
        public double magnitude { get; set; }

        public GPTVector2D()
        {
        }

        public GPTVector2D(double? x, double? y, double? mag, double? rad)
        {
            xx = x == null ? double.NaN : (double)x;
            yy = y == null ? double.NaN : (double)y;
            magnitude = mag == null ? double.NaN : (double)mag;
            theta_in_radians = rad == null ? double.NaN : (double)rad;
        }


        public GPTVector2D Add(GPTVector2D v1, GPTVector2D v2)
        {
            return new GPTVector2D(v1.xx + v2.xx, v1.yy + v2.yy, null, null);
        }

        public GPTVector2D Sub(GPTVector2D v1, GPTVector2D v2)
        {
            return new GPTVector2D(v1.xx - v2.xx, v1.yy - v2.yy, null, null);
        }

        public GPTVector2D Mul(GPTVector2D v1, double scalar)
        {
            return new GPTVector2D(v1.xx * scalar, v1.yy * scalar, null, null);
        }

        public GPTVector2D Div(GPTVector2D v1, double scalar)
        {
            if (scalar == 0)
                throw new DivideByZeroException("Cannot divide by zero.");

            return new GPTVector2D(v1.xx / scalar, v1.yy / scalar, null, null);
        }

        public double Dot(GPTVector2D v1, GPTVector2D v2)
        {
            return v1.xx * v2.xx + v1.yy * v2.yy;
        }

        public GPTVector2D Unit()
        {
            double len = Math.Sqrt(xx * xx + yy * yy);

            if (len == 0)
                throw new DivideByZeroException("Cannot normalize a zero vector.");

            return new GPTVector2D(xx / len, yy / len, 1.0, null);
        }

        public GPTVector2D Reverse()
        {
            return new GPTVector2D(100.0d - xx, 100.0d - yy, null, null);
        }

        public GPTVector2D ReverseUnit()
        {
            return new GPTVector2D(-xx, -yy, null, null);
        }

        public double ToRadians(double angle)
        {
            double res = angle * (Math.PI / 180.0d);

            return res;
        }

        public double ToDegrees(GPTVector2D v1)
        {
            double res = v1.theta_in_radians * (180.0d / Math.PI);

            return res;
        }

        public GPTVector2D ToPolar(GPTVector2D v1)
        {
            /*
             * r = √ ( x2 + y2 )
             * θ = tan^-1 ( y / x )
             * 
             * θ = arctan ( y / x )
             * 
             * in degrees
             * double theta1 = Math.Atan2(y, x) * 180.0 / Math.PI;
             * in radians
             * double theta2 = Math.Atan(y / x);
             * */

            double r = Math.Sqrt(v1.xx * v1.xx + v1.yy * v1.yy);
            double theta = Math.Atan2(v1.yy, v1.xx);

            v1.magnitude = r;
            v1.theta_in_radians = theta;

            return v1;
        }

        public GPTVector2D ToCart(GPTVector2D v1)
        {
            /*
             * x = r × cos( θ )
             * y = r × sin( θ )
             * */

            double x = v1.magnitude * Math.Cos(v1.theta_in_radians);
            double y = v1.magnitude * Math.Sin(v1.theta_in_radians);

            v1.xx = x;
            v1.yy = y;

            return v1;
        }
    }
}

File: Awesome.AI.Python/Common/GPTVector2D.py
from __future__ import annotations
import math


class GPTVector2D:
    def __init__(self, x=0.0, y=0.0, magnitude=None, radians=None):
        self.xx = math.nan if x is None else float(x)
        self.yy = math.nan if y is None else float(y)
        self.magnitude = math.nan if magnitude is None else float(magnitude)
        self.theta_in_radians = math.nan if radians is None else float(radians)

    @property
    def theta_in_degrees(self) -> float:
        return self.ToDegrees(self)

    def Add(self, v1, v2): return GPTVector2D(v1.xx + v2.xx, v1.yy + v2.yy, None, None)
    def Sub(self, v1, v2): return GPTVector2D(v1.xx - v2.xx, v1.yy - v2.yy, None, None)
    def Mul(self, v1, scalar): return GPTVector2D(v1.xx * scalar, v1.yy * scalar, None, None)
    def Div(self, v1, scalar):
        if scalar == 0: raise ZeroDivisionError("Cannot divide by zero.")
        return GPTVector2D(v1.xx / scalar, v1.yy / scalar, None, None)
    def Dot(self, v1, v2): return v1.xx * v2.xx + v1.yy * v2.yy
    def Unit(self):
        length = math.hypot(self.xx, self.yy)
        if length == 0: raise ZeroDivisionError("Cannot normalize a zero vector.")
        return GPTVector2D(self.xx / length, self.yy / length, 1.0, None)
    def Reverse(self): return GPTVector2D(100.0 - self.xx, 100.0 - self.yy, None, None)
    def ReverseUnit(self): return GPTVector2D(-self.xx, -self.yy, None, None)
    def ToRadians(self, angle): return math.radians(angle)
    def ToDegrees(self, v1): return math.degrees(v1.theta_in_radians)
    def ToPolar(self, v1): return GPTVector2D(v1.xx, v1.yy, math.hypot(v1.xx, v1.yy), math.atan2(v1.yy, v1.xx))
    def ToCart(self, v1): return GPTVector2D(v1.magnitude * math.cos(v1.theta_in_radians), v1.magnitude * math.sin(v1.theta_in_radians), v1.magnitude, v1.theta_in_radians)
